Passage à zmarkdown 9.1

Situphen opened the pull request:
Passage à zmarkdown 9.1

**QA :**

- `make zmd-install && make zmd-start && make run-back`
- Vérifier que le markdown n'est pas cassé
- Si aucune régression évidente n'est trouvée, merger

Situphen committed:
Passage à zmarkdown 9.1

StaloneLab commented:
Alors, je suis censé faire un PR ici, il y a des changements à faire côté `zds-site` pour profiter pleinement de la version (`linenostart` et `hl_lines` en particulier). Bien sûr, c'est une mineure, elle peut donc être déployée telle quelle, sans casser quoi que ce soit (par rapport à la 9.0, qu'on a pas testé par ailleurs). A voir comment on souhaite organiser les MàJ.

StaloneLab commented:
Par ailleurs, a-t-on toujours besoin de `pm2` alors qu'on utilise la version incluse dans le `node_modules` ?

Situphen committed:
Suppression de pm2 déjà installé via le paquet zmarkdown

Situphen commented:
> Alors, je suis censé faire un PR ici, il y a des changements à faire côté `zds-site` pour profiter pleinement de la version (`linenostart` et `hl_lines` en particulier). Bien sûr, c'est une mineure, elle peut donc être déployée telle quelle, sans casser quoi que ce soit (par rapport à la 9.0, qu'on a pas testé par ailleurs). A voir comment on souhaite organiser les MàJ.

Je propose : cette PR reste telle quelle pour l'instant et on la supprimera quand ta PR sortira si elle n'a pas été mergée entre temps.

> Par ailleurs, a-t-on toujours besoin de `pm2` alors qu'on utilise la version incluse dans le `node_modules` ?

Bien vu ! J'ai vérifié et effectivement c'est inutile car c'est déjà installé par le paquet `zmarkdown` donc je l'ai enlevé.

AmauryCarrade commented:
Je reprends cette PR et vais étudier l'intégration de `linenostart` et `hl_lines` (ne pas fermer, donc !).

AmauryCarrade committed:
Support pour `hl_lines` et `linenostart`.

- Ajoute le support pour `linenostart` en reprenant essentiellement le fonctionnement
  proposé par le ticket #5846, et en conservant la compatibilité avec le HTML généré
  par d'anciennes version de ZMD (avant la 9.1).
- Ajoute le support pour `hl_lines`, à l'aide d'un pseudo-élément attaché au numéro
  de ligne et étendu sur toute la largeur du bloc, sous le code. Node : cette
  implémentation ne fonctionne pas pour les langages qui nécessitent un fond de
  couleur (à ma connaissance, ce n'est que le cas de `console`).

AmauryCarrade commented:
> Je reprend cette PR et vais étudier l'intégration de `linenostart` et `hl_lines` (ne pas fermer, donc !).

Finalement ce fut plus simple que prévu (cf. dernier commit). Le support est fonctionnel.

En effet le surlignage est implémenté en utilisant un pseudo-élément attaché sur le numéro de la ligne (car seul ce dernier a l'info de la ligne surlignée) et placé _sous_ le code, afin que ce dernier reste sélectionnable.

D'un point de vue accessibilité : l'info que la ligne est surlignée n'est pas du tout disponible pour les lecteurs d'écran, mais je ne vois pas trop comment la donner (les numéros de lignes sont pas disponibles non plus je pense, mais je n'ai pas testé). Le contraste est bon sauf pour les commentaires où il est un peu faible, mais ce problème existe aussi sans surlignage : il faudrait donc ajuster leur couleur tout court.

AmauryCarrade committed:
ZMD 9.1.2 + amélioration des blocs de console.

- Passage à ZMD 1.9.2 qui ajoute la possibilité de styliser en fonction du
  langage, via une classe CSS sur l'élément parent.
- Amélioration des éléments console : intégration dans le fond noir des numéros
  de ligne, support du surlignement, et amélioration des couleurs.

AmauryCarrade commented:
Le dernier commit ajoute le support complet pour les extrait de console, en plus du reste. On passe à zmd 1.9.2.

Les consoles sont également mieux intégrées, et l'accessibilité a été vérifiée et améliorée.

![Aperçu de la nouvelle intégration et des surlignements & décalages de lignes.](https://user-images.githubusercontent.com/1417570/86612045-d0b61f80-bfaf-11ea-9ca4-516f0b14410f.png)

Au passage, :heavy_check_mark: QA initiale OK, je n'ai pas vu d'éléments cassés par le MD. Mais re-vérifier en faisant la QA ci-dessous ne peut pas faire de mal.

### Contrôle qualité

1. **Avant de mettre à jour ZMD** (donc avec l'ancienne version), lancer le serveur et créer un message, ou quelque chose, qui contienne un code de type `console` : 
    ``````markdown
    ```console
    $ sl
    * tchou tchou *
    ```
    ``````
2. Mettre à jour ZMD et relancer le serveur (`make zmd-install && make zmd-stop && make zmd-start && make run`).
3. Dans un autre message que celui de la première étape (qui doit resté intouché et non-mis-à-jour), insérer du code avec des numéros de lignes ne débutant pas à 1 et des lignes surlignées. Insérer également un code de type console avec ces fonctionnalités.
4. Vérifier : 
    - que les bonnes lignes sont sélectionnées, et que les numéros de ligne commencent à la bonne valeur ;
    - que le bloc de code console est correctement affiché (avec tout sur fond noir, y compris les numéros de ligne), et que les surlignages et numéros décalés marchent ;
    - que les blocs de code dans les messages utilisant l'ancienne version de ZMD (le message créé avant la mise à jour) sont toujours correctement affichés.

Exemple de code pour tester, trouvé dans le hasard : 

``````markdown
```c linenostart=229 hl_lines=234,237
char *
__initstate (unsigned int seed, char *arg_state, size_t n)
{
  int32_t *ostate;
  int ret;
  __libc_lock_lock (lock);
  ostate = &unsafe_state.state[-1];
  ret = __initstate_r (seed, arg_state, n, &unsafe_state);
  __libc_lock_unlock (lock);
  return ret == -1 ? NULL : (char *) ostate;
}
```

```console hl_lines=2
$ sl
* tchou tchou *
```
``````

philippemilink commented:
Rapport de QA
---------------

(j'allais dire que c'est toujours un plaisir de faire des QAs avec un protocole si bien détaillé, mais tu as oublié de dire qu'il fallait faire `make build-front` 😉  )

Presque OK.

Pour les lignes surlignées et le début de la numérotation différent de 1: pas de problème.

C'est pour l'extrait de console le problème: l'apparence n'est pas la même pour les messages écrits avant le passage à ZMD 9.1:
![image](https://user-images.githubusercontent.com/5911232/86836763-426c9700-c09e-11ea-8be3-2fe7812a3cf3.png)

En plus, il faudra penser à rebase.

AmauryCarrade commented:
C'est normal car il n'est pas possible d'appliquer le nouveau style sur les anciens rendus : ces derniers manquent d'une classe CSS pour ça, ajoutée par zmd 9.1.2, et donc absente des messages rendus avec une version antérieure.

Les anciens rendus doivent rendre _correctement_ malgré tout (d'où cette partie de la QA), mais pas être _identiques_.

Demain je fais le rebase (et j'ajouterai aussi quelques _protips_ Markdown liés à ces nouvelles fonctionnalités de zmd, mais ça ne demandera pas de QA autre que Travis).

> (j'allais dire que c'est toujours un plaisir de faire des QAs avec un protocole si bien détaillé, mais tu as oublié de dire qu'il fallait faire `make build-front` :wink: )

Hum, ce n'est normalement pas nécessaire étant donné que `make run` reconstruit le _front_… C'est pour ça que je ne l'ai pas précisé.

artragis commented:
@philippemilink tu peux considérer que le rendu que tu proposes est celui attendu. Merci de l'avoir noté.

@AmauryCarrade je t'invite à noter quelque part dans le projet (commentaire du CSS) que ce genre de comportement est en effet attendu comme ça on tombera pas des nues dans 9 mois et 27 jours et que quelqu'un viendra nous dire que ça fait ça alors qu'il s'attend à autre chose.

AmauryCarrade committed:
Subtile amélioration des couleurs (palette) et commentaires.

- Modification mineure des couleurs afin d'anticiper la future palette.
- Ajout de commentaires pour expliciter le fait qu'il y a deux styles de 
console.

AmauryCarrade committed:
Ajout d'astuces Markdown relatives aux nouvelles possibilités de zmd 9.1

AmauryCarrade commented:
> @AmauryCarrade je t'invite à noter quelque part dans le projet (commentaire du CSS) que ce genre de comportement est en effet attendu comme ça on tombera pas des nues dans 9 mois et 27 jours et que quelqu'un viendra nous dire que ça fait ça alors qu'il s'attend à autre chose.

@artragis C'est fait ! J'ai également ajouté des astuces MD et mis la branche à jour par rapport à `dev`. Quand Travis sera contente, ce sera fusionnable, sauf avis divergents sur les astuces Markdown, ou remarques sur le code.

artragis commented on assets/scss/base/_source-code.scss:
```diff
@@ -90,25 +94,32 @@
 
+  // This applies only to console codes rendered with zmd 9.1.2+, as previous
+  // versions were missing the possibility to style them as full console blocks.
+  // Old blocs only have a bkack background on the content, not the line numbers,
```
-bkak + black

artragis commented on assets/scss/base/_source-code.scss:
```diff
@@ -90,25 +94,32 @@
       span {
         &:after , &.hll:after{
-          color: white;  // TODO $white
+          color: hsl(0, 0%, 98%);  // TODO $white
```
pourquoi ne pas mettre une variable ici?

AmauryCarrade replied:
La variable existe déjà dans #5835, d'où le commentaire TODO, mais je ne jugeais pas pertinent de créer la variable juste pour ça. Cela dit je ne m'y oppose pas, si vous trouvez que ce serait mieux.

artragis commented on assets/scss/base/_source-code.scss:
```diff
@@ -90,25 +94,32 @@
+    // Required for highlight to work
     pre code.language-console {
-      background-color: transparent !important;
+      background-color: transparent;
```
gg

artragis commented on assets/scss/base/_source-code.scss:
```diff
@@ -90,25 +94,32 @@
 }
 
+// The code part of the console codes (shared for legacy and new versions,
+// the background-color being overrided for new versions).
```
overridden

AmauryCarrade committed:
Typos (merci artragis)

StaloneLab commented on assets/js/markdown-help.js:
````diff
@@ -24,7 +24,10 @@
     'Le langage d’un bloc de code peut être spécifié après les <code>```</code> ouvrants. La liste des langages supportés <a href="' + linkToPygments + '">est disponible ici</a>.',
     'Vous pouvez <a href="' + linkToMathsTutorial + '">écrire des formules mathématiques</a> en encadrant ces dernières du signe dollar <code>$</code>.',
-    'Pour ajouter une image, vous pouvez simplement la glisser-déposer depuis votre explorateur.'
+    'Pour ajouter une image, vous pouvez simplement la glisser-déposer depuis votre explorateur.',
+    'vous pouvez préciser à quel numéro commencent les lignes avec cette syntaxe : <code>```python linenostart=42</code> au début d\'un bloc de code. Pratique pour faire coïncider les numéros de ligne à une erreur, par exemple.',
+    'vous pouvez surligner des lignes avec cette syntaxe : <code>```rust hl_lines=2,4-7</code> au début d\'un bloc de code.',
+    'vous pouvez à la fois choisir à quel numéro démarrent les lignes et en surligner avec <code>```lisp linenostart=244 hl_lines=247,252</code> au début d\'un bloc de code.'
````
Pas sûr de moi, mais ces astuces ne devraient-elles pas commencer par des majuscules comme les autres ?

AmauryCarrade replied:
J'ai tout passé en minuscules sur #5835 parce que le rendu est typographiquement meilleur : 

> **Astuce :** Pour ajouter une image, vous pouvez…
> **Astuce :** pour ajouter une image, vous pouvez…

Mais si on veut de la cohérence, je peux remettre les majuscules ici et ne les retirer que là bas. Ou pas du tout si vous tenez à l'incohérence typographique.

StaloneLab replied:
Ben du coup je serais d'avis de les retirer partout

StaloneLab replied:
Ah, je viens de comprendre que tu l'avais déjà fait ailleurs, désolé du coup

AmauryCarrade replied:
Je pourrais les retirer ici aussi en soit (en déplaçant la modification), dans les deux cas c'est pas directement lié à la PR

AmauryCarrade committed:
Cohérence.

AmauryCarrade commented:
Si ce n'est l'histoire de variable on est bon. J'attends juste confirmation pour cette dernière.
